effectful/handlers/llm/synthesis.py

```diff
@@ -19,82 +26,425 @@ def __init__(self, message, code=None):
-        code = code_content.group(1)
+    def __init__(self, type_check: bool = False):
+        """Initialize the program synthesis handler.
 
-        try:
-            module_ast = ast.parse(code)
-        except SyntaxError as exc:
-            raise SynthesisError("failed to parse", content) from exc
+        Args:
+            type_check: Whether to run mypy to verify the generated code.
+                        Even with constrained decoding, this can catch errors
+                        in the function body implementation.
+        """
+        self.type_check = type_check
+
+    def _build_function(
+        self,
+        result: SynthesizedFunction,
+        callable_type: type,
+        referenced_types: set[type],
+    ) -> typing.Callable:
+        """Build and execute a function from the structured synthesis result.
+
+        Args:
+            result: The structured output from the LLM
+            callable_type: The expected Callable type (e.g., Callable[[str], int])
+            referenced_types: Set of types referenced in the signature
 
-        if not isinstance(module_ast, ast.Module):
-            raise SynthesisError("not a module", content)
+        Returns:
+            The synthesized callable function
+        """
+        # Build the function with prescribed types and LLM-provided names
+        param_sig = _format_param_signature(callable_type, result.param_names)
+        return_type = _format_return_type(callable_type)
+        func_name = result.function_name
 
-        last_decl = module_ast.body[-1]
-        if not isinstance(last_decl, ast.FunctionDef):
-            raise SynthesisError("last definition not a function", content)
+        # Ensure body is properly indented
+        body = result.body
+        if not body.startswith("    ") and not body.startswith("\t"):
+            # Indent the body if not already indented
+            body = textwrap.indent(body, "    ")
 
-        source_code = textwrap.dedent(code)
+        # Construct the full function code
+        code = f"def {func_name}({param_sig}) -> {return_type}:\n{body}"
+
+        # Register in linecache for better tracebacks
+        source_code = code
         lines = code.splitlines(keepends=True)
         filename = f"<generated-{hash(code)}>"
-
-        # register into linecache
         linecache.cache[filename] = (len(source_code), None, lines, filename)
 
-        # TODO: assert callable type compatibility
+        # Optional mypy type checking
+        if self.type_check:
+            success, error_msg = run_mypy_check(code, referenced_types)
+            if not success:
+                raise SynthesisError(f"Type check failed:\n{error_msg}", code)
+
+        # Build globals dict by importing types from their original modules
         gs: dict = {}
+        for typ in referenced_types:
+            module = inspect.getmodule(typ)
+            if module is not None:
+                gs[typ.__name__] = typ
```

Comment:
Why do we do this? Presumably, the code already contains the imports we need because we built them for mypy. We should pull out the part of `run_mypy_check` that adds imports and do that unconditionally so we don't duplicate the behavior.